builtin/logical/ssh/util.go

```diff
@@ -0,0 +1,206 @@
+package ssh
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// Transfers the file  to the target machine by establishing an SSH
+// session with the target. Uses the public key authentication method
+// and hence the parameter 'key' takes in the private key. The fileName
+// parameter takes an absolute path.
+func uploadPublicKeyScp(publicKey, username, ip, port, key string) error {
+	dynamicPublicKeyFileName := fmt.Sprintf("vault_ssh_%s_%s.pub", username, ip)
+	session, err := createSSHPublicKeysSession(username, ip, port, key)
+	if err != nil {
+		return err
+	}
+	if session == nil {
+		return fmt.Errorf("invalid session object")
+	}
+	defer session.Close()
+	go func() {
+		w, _ := session.StdinPipe()
+		fmt.Fprintln(w, "C0644", len(publicKey), dynamicPublicKeyFileName)
+		io.Copy(w, strings.NewReader(publicKey))
+		fmt.Fprint(w, "\x00")
+		w.Close()
+	}()
+	err = session.Run(fmt.Sprintf("scp -vt %s", dynamicPublicKeyFileName))
+	if err != nil {
+		return fmt.Errorf("public key upload failed")
+	}
+	return nil
+}
+
+// Creates a SSH session object which can be used to run commands
+// in the target machine. The session will use public key authentication
+// method with port 22.
+func createSSHPublicKeysSession(username, ipAddr, port, hostKey string) (*ssh.Session, error) {
+	if username == "" {
+		return nil, fmt.Errorf("missing username")
+	}
+	if ipAddr == "" {
+		return nil, fmt.Errorf("missing ip address")
+	}
+	if hostKey == "" {
+		return nil, fmt.Errorf("missing host key")
+	}
+	signer, err := ssh.ParsePrivateKey([]byte(hostKey))
+	if err != nil {
+		return nil, fmt.Errorf("parsing Private Key failed: %s", err)
+	}
+
+	config := &ssh.ClientConfig{
+		User: username,
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(signer),
+		},
+	}
+
+	client, err := ssh.Dial("tcp", fmt.Sprintf("%s:%s", ipAddr, port), config)
+	if err != nil {
+		return nil, err
+	}
+	if client == nil {
+		return nil, fmt.Errorf("invalid client object: %s", err)
+	}
+
+	session, err := client.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	return session, nil
+}
+
+// Creates a new RSA key pair with key length of 2048.
+// The private key will be of pem format and the public key will be
+// of OpenSSH format.
+func generateRSAKeys() (publicKeyRsa string, privateKeyRsa string, err error) {
```

Comment:
Can we make the 2048 configurable instead of hard coded? A property of the role perhaps?